### PR TITLE
feat: add "run" command to the client interface inorder to allow the …

### PR DESCRIPTION
**Pull request by khanayan123**
…running of scripts against drivers

## Proposed changes

To create a generic script running CLI command that allows commands of the form appium driver run <driverName> <scriptName>.

## Types of changes

What types of changes does your code introduce to Appium?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/appium/appium/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla.js.foundation/appium/appium)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments
I have not added necessary documentation yet. Wanted to get some feedback on the PR first. The E2E tests in this PR require the merging of this branch: https://github.com/appium/appium-fake-driver/pull/76 
in order to work.

### CODE REVIEW

**Review by jlipps — CHANGES_REQUESTED** [review]
looking really good! i made some larger comments and will look at tests once we have those other things hammered out.

File: lib/cli/args.js [hunk 1]
```diff
@@ -415,11 +415,27 @@ function makeUpdateArgs (type) {
   ];
 }
 
+function makeRunArgs (type) {
+  return [
+    ...globalExtensionArgs,
+    [[type], {
+      type: 'str',
+      help: `Name of the ${type} to run a script from, for example: ` + type === DRIVER_TYPE ? DRIVER_EXAMPLE : PLUGIN_EXAMPLE,
```
Comment (jlipps) [inline lib/cli/args.js:423]: Break this into different lines for readability as above, I think our max length is around 100 chars

File: lib/cli/extension-command.js [hunk 2]
```diff
@@ -486,4 +487,49 @@ export default class ExtensionCommand {
...
+  /**
+   * Runs a script cached inside the "scripts" field under "appium"
+   * inside of the driver/plugins "package.json" file
+   *
+   * @param {string} ext - name of the extension to run a script from
+   * @param {string} scriptName - name of the script to run
+   */
+  async run ({ext, scriptName}) {
+    if (!_.has(this.config.installedExtensions, ext)) {
+      const msg = `please install the ${this.type} first`;
```
Comment (jlipps) [inline lib/cli/extension-command.js:500]: no need to define the string and throw in a separate line. just put the string literal in the Error constructor

File: lib/cli/extension-command.js [hunk 3]
```diff
@@ -486,4 +487,49 @@ export default class ExtensionCommand {
...
+    const extConfig = this.config.installedExtensions[ext];
+
+    if (!_.has(extConfig, 'scripts')) {
+      throw new Error(`The ${this.type} named '${ext}' does not contain the ` +
+                      `"scripts" field underneath the "appium" field in its package.json`);
+    }
+
+    const extScripts = extConfig.scripts;
+
+    if (!(scriptName in extScripts)) {
```
Comment (jlipps) [inline lib/cli/extension-command.js:513]: use `_.has` as in the other places

File: lib/cli/extension-command.js [hunk 4]
```diff
@@ -486,4 +487,49 @@ export default class ExtensionCommand {
...
+      throw new Error(`The ${this.type} named '${ext}' does not support the script: '${scriptName}'`);
+    }
+
+    const runner = new SubProcess(`node`, [extScripts[scriptName]],
+                                  {cwd: this.config.getExtensionRequirePath(ext)});
+
+    runner.on('stream-line', (line) => {
+      log(this.isJsonOutput, line);
```
Comment (jlipps) [inline lib/cli/extension-command.js:521]: here we can also keep an array of output lines that we can use in the JSON return

File: lib/cli/extension-command.js [hunk 5]
```diff
@@ -486,4 +487,49 @@ export default class ExtensionCommand {
...
+    });
+
+    await runner.start();
+
+    try {
+      await runner.join();
+      log(this.isJsonOutput, `${scriptName} successfuly ran`.green);
+      return {scriptName};
```
Comment (jlipps) [inline lib/cli/extension-command.js:529]: it would be more useful to return the output than just the script name, as an array of output lines that we capture above

File: lib/cli/extension-command.js [hunk 6]
```diff
@@ -486,4 +487,49 @@ export default class ExtensionCommand {
...
+    } catch (err) {
+      log(this.isJsonOutput, `Encountered an error when running the script: ${err.message}`.red);
+      return {error: err.message};
```
Comment (jlipps) [inline lib/cli/extension-command.js:532]: in addition to the error message, it would be nice to include any output in the JSON response, since the error message might be something generic like 'exited with code 1'.

**Review by jlipps — CHANGES_REQUESTED** [review]

File: test/cli-e2e-specs.js [hunk 7]
```diff
@@ -142,4 +143,87 @@ describe('Driver CLI', function () {
...
+    it('should run a valid plugin, valid script, and result in success', async function () {
+      await clear();
+      const pluginName = 'fake';
+      const scriptName = 'fake-success';
+      await run('install', [fakePluginDir, '--source', 'local', '--json']);
+      const out = JSON.parse(await run('run', [pluginName, scriptName, '--json']));
+      out.should.not.include.key('error');
+    });
+    it('should run a valid plugin, valid error prone script, and return error in json', async function () {
+      await clear();
```
Comment (jlipps) [inline test/cli-e2e-specs.js:212]: in this and the next test, we theoretically don't need to clear and install, right? what about just clearing and installing once in a `before` for this whole describe?

**Review by jlipps — APPROVED** [review]

**Comment by jlipps** [comment]
also, looks like CI is failing. any ideas?

**Comment by ayvnkhan** [comment]
> also, looks like CI is failing. any ideas?
I'd assume it is because this branch https://github.com/appium/appium-plugins/pull/7 hasn't been merged yet.

**Comment by jlipps** [comment]
now these changes need to be applied to the new appium directory location in the 2.0 branch

File: packages/appium/lib/cli/extension-command.js [hunk 8]
```diff
@@ -486,4 +487,51 @@ export default class ExtensionCommand {
...
+    const extConfig = this.config.installedExtensions[ext];
+
+    if (!_.has(extConfig, 'scripts')) {
+      throw new Error(`The ${this.type} named '${ext}' does not contain the ` +
+                      `"scripts" field underneath the "appium" field in its package.json`);
+    }
+
+    const extScripts = extConfig.scripts;
+
+    if (!_.has(extScripts, scriptName)) {
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:512]: I'd also check if it is a plain object here

File: packages/appium/lib/cli/extension-command.js [hunk 9]
```diff
@@ -486,4 +487,51 @@ export default class ExtensionCommand {
...
     await this.config.write();
   }
-}
+
+  /**
+   * Runs a script cached inside the "scripts" field under "appium"
+   * inside of the driver/plugins "package.json" file
+   *
+   * @param {string} ext - name of the extension to run a script from
+   * @param {string} scriptName - name of the script to run
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:496]: missing the description of the returned result and of possible thrown exceptions

File: packages/appium/lib/cli/extension-command.js [hunk 10]
```diff
@@ -486,4 +487,51 @@ export default class ExtensionCommand {
...
+    const extScripts = extConfig.scripts;
+
+    if (!_.has(extScripts, scriptName)) {
+      throw new Error(`The ${this.type} named '${ext}' does not support the script: '${scriptName}'`);
+    }
+
+    const runner = new SubProcess(`node`, [extScripts[scriptName]],
+                                  {cwd: this.config.getExtensionRequirePath(ext)});
+
+    const output = [];
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:519]: I'm not quite sure we'd like to keep the whole output. It might create OOM in the worst case. I'd prefer to limit it's size and organize a ring buffer

File: packages/appium/lib/cli/extension-command.js [hunk 11]
```diff
@@ -486,4 +487,51 @@ export default class ExtensionCommand {
...
+
+    runner.on('stream-line', (line) => {
+      output.push(line);
+      log(this.isJsonOutput, line);
+    });
+
+    await runner.start();
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:526]: this might miserably fail if there was no output for more than 5 seconds

File: packages/appium/lib/cli/extension-command.js [hunk 12]
```diff
@@ -509,29 +519,39 @@ export default class ExtensionCommand {
 
     const extScripts = extConfig.scripts;
 
+    if (!_.isPlainObject(extScripts)) {
+      throw new Error(`"scripts" field must be a plain object`);
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:523]: I would also add the extension name here, because there might be multiple

File: packages/appium/lib/cli/extension-command.js [hunk 13]
```diff
@@ -509,29 +519,39 @@ export default class ExtensionCommand {
...
       output.push(line);
       log(this.isJsonOutput, line);
     });
 
-    await runner.start();
+    try {
+      // start and wait 5 seconds before throwing an error
+      await runner.start(null, 5 * 1000);
+    } catch (err) {
+      log(this.isJsonOutput, `Encountered an error when running the script: ${err.message}`.red);
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:544]: I would also add the script name here

File: packages/appium/lib/cli/utils.js [hunk 14]
```diff
@@ -52,9 +52,24 @@ async function spinWith (json, msg, fn) {
   }
 }
 
+let createRingBuffer = function (length) {
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/utils.js:55]: is there any particular reason why this must be an anonymous function?

File: packages/appium/lib/cli/utils.js [hunk 15]
```diff
@@ -52,9 +52,24 @@ async function spinWith (json, msg, fn) {
...
+
+  let pointer = 0, buffer = [];
+
+  return {
+    getBuf () {return buffer;},
+    get (key) {return buffer[key];},
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/utils.js:61]: I would call it index rather than key

File: packages/appium/lib/cli/utils.js [hunk 16]
```diff
@@ -52,9 +52,24 @@ async function spinWith (json, msg, fn) {
   }
 }
 
+let createRingBuffer = function (length) {
+
+  let pointer = 0, buffer = [];
+
+  return {
+    getBuf () {return buffer;},
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/utils.js:60]: is linter happy about it?

File: packages/appium/lib/cli/extension-command.js [hunk 17]
```diff
@@ -488,12 +488,22 @@ export default class ExtensionCommand {
     await this.config.write();
   }
 
+  /**
+   * @typedef RunOutput
+   * @property {string|undefined} error - error message if script ran unsuccessfuly, otherwise undefined
+   * @property {array} output - script output
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:494]: `{string[]} or {Array<string>}

File: packages/appium/lib/cli/extension-command.js [hunk 18]
```diff
@@ -488,12 +488,22 @@ export default class ExtensionCommand {
     await this.config.write();
   }
 
+  /**
+   * @typedef RunOutput
+   * @property {string|undefined} error - error message if script ran unsuccessfuly, otherwise undefined
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:493]: I usually use `{?string}`, but the current one is also ok

File: packages/appium/lib/cli/utils.js [hunk 19]
```diff
@@ -52,9 +52,24 @@ async function spinWith (json, msg, fn) {
   }
 }
 
+let createRingBuffer = function (length) {
+
+  let pointer = 0, buffer = [];
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/utils.js:57]: I prefer to define each variable in a separate row to improve readability

File: packages/appium/lib/cli/utils.js [hunk 20]
```diff
@@ -52,9 +52,24 @@ async function spinWith (json, msg, fn) {
...
+
+  return {
+    getBuf () {return buffer;},
+    get (key) {return buffer[key];},
+    push (item) {
+      buffer[pointer] = item;
+      pointer = (length + pointer + 1) % length;
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/utils.js:64]: I think it might be confusing if we start recording recent logs to the beginning of the buffer. I would rather shift the buffer to the left and delete the very first record in order to add the last one, so that the original order is preserved

File: packages/appium/lib/cli/extension-command.js [hunk 21]
```diff
@@ -486,4 +487,71 @@ export default class ExtensionCommand {
...
+    const output = createRingBuffer(50);
+
+    runner.on('stream-line', (line) => {
+      output.push(line);
+      log(this.isJsonOutput, line);
+    });
+
+    try {
+      // start and wait 5 seconds before throwing an error
+      await runner.start(null, 5 * 1000);
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:542]: I usually avoid creating magic number and put them to constants instead

File: packages/appium/lib/cli/extension-command.js [hunk 22]
```diff
@@ -486,4 +487,71 @@ export default class ExtensionCommand {
...
+
+    const output = createRingBuffer(50);
+
+    runner.on('stream-line', (line) => {
+      output.push(line);
+      log(this.isJsonOutput, line);
+    });
+
+    try {
+      // start and wait 5 seconds before throwing an error
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:541]: I'm actually not sure whether we need to wait for any output here at all. A script is not obliged to print anything. I would simply use `start(0)` and let it throw when `join` is called

File: packages/appium/lib/cli/extension-command.js [hunk 23]
```diff
@@ -491,7 +491,7 @@ export default class ExtensionCommand {
   /**
    * @typedef RunOutput
    * @property {?string|undefined} error - error message if script ran unsuccessfuly, otherwise undefined
-   * @property {array} output - script output
+   * @property {string[]|Array<string>} output - script output
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:494]: I meant one of these types annotations is ok, not both of them at once :)

File: packages/appium/lib/cli/utils.js [hunk 24]
```diff
@@ -52,9 +52,30 @@ async function spinWith (json, msg, fn) {
...
+    this.buffer = [];
+  }
+  getBuff () {
+    return this.buffer;
+  }
+  dequeue () {
+    this.buffer.shift();
+  }
+  enqueue (item) {
+    if (this.buffer.length === this.size) {
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/utils.js:68]: `>=` condition looks safer to me

File: packages/appium/lib/cli/extension-command.js [hunk 25]
```diff
@@ -486,4 +487,65 @@ export default class ExtensionCommand {
...
+
+    if (!_.isPlainObject(extScripts)) {
+      throw new Error(`The ${this.type} named '${ext}' "scripts" field must be a plain object`);
+    }
+
+    if (!_.has(extScripts, scriptName)) {
+      throw new Error(`The ${this.type} named '${ext}' does not support the script: '${scriptName}'`);
+    }
+
+    const runner = new SubProcess(`node`, [extScripts[scriptName]],
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:530]: nit: we usually format such arguments like: 
```
const runner = new SubProcess(`node`, [extScripts[scriptName]], {
  cwd: this.config.getExtensionRequirePath(ext)
});
```

**Review by mykola-mokhnach — APPROVED** [review]
had some minor comments

File: packages/appium/lib/cli/utils.js [hunk 26]
```diff
@@ -52,9 +52,30 @@ async function spinWith (json, msg, fn) {
   }
 }
 
+class RingBuffer {
+  constructor (size = 50) {
+    this.size = size;
+    this.pointer = 0;
```
Comment (jlipps) [inline packages/appium/lib/cli/utils.js:58]: `this.pointer` doesn't seem to be used anywhere. should we delete it?

File: packages/appium/test/cli-e2e-specs.js [hunk 27]
```diff
@@ -15,6 +15,7 @@ const executable = path.join(cwd, 'packages', 'appium', 'build', 'lib', 'main.js
 
 describe('Driver CLI', function () {
   let appiumHome;
+  const localFakeDriverPath = path.resolve(cwd, 'packages', 'fake-driver');
```
Comment (jlipps) [inline packages/appium/test/cli-e2e-specs.js:18]: this resolution strategy will work only if running tests from the main appium directory. what if we're running them inside packages/appium? It might be better to do something like `path.resolve(__dirname, '..', '..', '..', 'fake-driver')` (with an extra '..' in the list because the transpiled code will run from `packages/appium/build`, not `packages/appium`)

Comment (jlipps) [reply packages/appium/test/cli-e2e-specs.js:18]: i'm ok if we leave this as it is right now, i understand you were just following the existing strategy.

File: packages/appium/test/cli-e2e-specs.js [hunk 28]
```diff
@@ -146,4 +146,81 @@ describe('Driver CLI', function () {
...
+      const driverName = 'foo';
+      await chai.expect(run('run', [driverName, 'bar', '--json'])).to.eventually.be.rejectedWith(Error);
+    });
+  });
+});
+
+describe('Plugin CLI', function () {
+  let appiumHome;
+  const fakePluginDir = path.resolve(cwd, 'node_modules', '@appium', 'fake-plugin');
+  before(async function () {
```
Comment (jlipps) [inline packages/appium/test/cli-e2e-specs.js:180]: we duplicate all this before, after, run, and clear logic. seems like we could easily make them shared helper functions? (and there's probably no downside to sharing `appiumHome` between the driver and plugin tests here.)

**Review by jlipps — CHANGES_REQUESTED** [review]
everything looks great! just a couple thoughts as well as a request to deduplicate some of the test code.

**Review by jlipps — APPROVED** [review]

**Comment by jlipps** [comment]
looks like you're out of CI probation now at least :-)

**Comment by ayvnkhan** [comment]
> looks like you're out of CI probation now at least :-)

haha thank god

File: packages/appium/lib/cli/extension-command.js [hunk 29]
```diff
@@ -486,4 +487,51 @@ export default class ExtensionCommand {
...
+                      `"scripts" field underneath the "appium" field in its package.json`);
+    }
+
+    const extScripts = extConfig.scripts;
+
+    if (!_.has(extScripts, scriptName)) {
+      throw new Error(`The ${this.type} named '${ext}' does not support the script: '${scriptName}'`);
+    }
+
+    const runner = new SubProcess(`node`, [extScripts[scriptName]],
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/extension-command.js:516]: Consider using `process.execPath` instead of `node` as the latter won't work on Windows. See https://stackoverflow.com/questions/49934315/how-to-get-path-of-nodejs-executable-on-runtime

File: packages/appium/lib/cli/extension-command.js (hunk 11 above)
Comment (mykola-mokhnach) [reply packages/appium/lib/cli/extension-command.js:526]: still not resolved. Please pass a single `0` argument to the `start` call to avoid any checks

File: packages/appium/lib/cli/extension-command.js (hunk 9 above)
Comment (mykola-mokhnach) [reply packages/appium/lib/cli/extension-command.js:496]: still not resolved. Is the method supposed to return a string a an array of strings?

File: packages/appium/lib/cli/parser.js [hunk 30]
```diff
@@ -98,8 +98,16 @@ function addExtensionsToParser (sharedArgs, subParsers, debug = false) {
...
       {command: 'update', args: extensionArgs[type].update,
        help: `Update installed ${type}s to the latest version`},
+      {command: 'run', args: extensionArgs[type].run,
+       help: `Run a script (cached inside the ${type}s package.json under the ` +
+             `“scripts” field inside the “appium” field) from an installed ${type}`}
     ];
 
+    if (type === DRIVER_TYPE) {
+      parserSpecs.push({command: 'run', args: extensionArgs[type].run,
+                        help: `Run a script (cached inside the ${type}s package.json under the “scripts” field inside the “appium” field) against an installed ${type}`});
```
Comment (mykola-mokhnach) [inline packages/appium/lib/cli/parser.js:108]: this line is too long

File: packages/appium/lib/cli/extension-command.js (hunk 9 above)
Comment (ayvnkhan) [reply packages/appium/lib/cli/extension-command.js:496]: Seems like I must've messed up during git rebase, let me try again these comments should have been resolved

File: packages/appium/lib/cli/parser.js [hunk 31]
```diff
@@ -98,6 +98,9 @@ function addExtensionsToParser (sharedArgs, subParsers, debug = false) {
        help: `Uninstall a ${type}`},
       {command: 'update', args: extensionArgs[type].update,
        help: `Update installed ${type}s to the latest version`},
+      {command: 'run', args: extensionArgs[type].run,
+       help: `Run a script (cached inside the ${type}s package.json under the ` +
```
Comment (jlipps) [inline packages/appium/lib/cli/parser.js:102]: I would say `defined` rather than `cached`. I'm not sure what cached would mean here.

**Review by jlipps — APPROVED** [review]

File: packages/appium/lib/cli/parser.js (hunk 31 above)
Comment (jlipps) [reply packages/appium/lib/cli/parser.js:102]: also, `${type}'s`

**Review by jlipps — APPROVED** [review]

**Comment by jlipps** [comment]
\o/ nice work!
